pkg/declarative/v2/reconciler.go

```diff
@@ -0,0 +1,382 @@
+		obj.SetStatus(status.WithState(StateProcessing).WithOperation("switched to processing"))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	spec, err := r.Spec(ctx, obj)
+	if err != nil {
+		r.Event(obj, "Warning", "Spec", err.Error())
+		obj.SetStatus(status.WithState(StateError).WithErr(err))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	skr, err := r.initializeClient(ctx, obj, spec)
+	if err != nil {
+		r.Event(obj, "Warning", "ClientInitialization", err.Error())
+		obj.SetStatus(status.WithState(StateError).WithErr(err))
+		return r.ssaStatus(ctx, obj)
+	}
+	resourceConverter := NewResourceConverter(skr, r.Namespace)
+
+	var renderer Renderer
+
+	switch spec.Mode {
+	case RenderModeHelm:
+		renderer = NewHelmRenderer(spec, skr, r.Options)
+		renderer = WrapWithRendererCache(renderer, spec, r.Options)
+	case RenderModeKustomize:
+		renderer = NewKustomizeRenderer(spec, r.Options)
+		renderer = WrapWithRendererCache(renderer, spec, r.Options)
+	case RenderModeRaw:
+		renderer = NewRawRenderer(spec, r.Options)
+	}
+
+	if err := renderer.Initialize(obj); err != nil {
+		return r.ssaStatus(ctx, obj)
+	}
+
+	if err := renderer.EnsurePrerequisites(ctx, obj); err != nil {
+		return r.ssaStatus(ctx, obj)
+	}
+
+	var target, current kube.ResourceList
+
+	if obj.GetDeletionTimestamp().IsZero() {
+		var targetResources *types.ManifestResources
+
+		manifest, err := renderer.Render(ctx, obj)
+		if err != nil {
+			return r.ssaStatus(ctx, obj)
+		}
+
+		targetResources, err = util.ParseManifestStringToObjects(string(manifest))
+		if err != nil {
+			r.Event(obj, "Warning", "ManifestParsing", err.Error())
+			obj.SetStatus(status.WithState(StateError).WithErr(err))
+			return r.ssaStatus(ctx, obj)
+		}
+
+		targetResources.Items = append(targetResources.Items, spec.TargetResources...)
+
+		for _, transform := range r.PostRenderTransforms {
+			if err := transform(ctx, obj, targetResources); err != nil {
+				r.Event(obj, "Warning", "PostRenderTransform", err.Error())
+				obj.SetStatus(status.WithState(StateError).WithErr(err))
+				return r.ssaStatus(ctx, obj)
+			}
+		}
+
+		target, err = resourceConverter.ConvertResourcesFromManifest(targetResources)
+		if err != nil {
+			r.Event(obj, "Warning", "TargetResourceParsing", err.Error())
+			resourceCondition.Status = metav1.ConditionFalse
+			meta.SetStatusCondition(&status.Conditions, resourceCondition)
+			obj.SetStatus(status.WithState(StateError).WithErr(err))
+			return r.ssaStatus(ctx, obj)
+		}
+	} else {
+		target = kube.ResourceList{}
+	}
+
+	if current, err = resourceConverter.ConvertStatusToResources(status); err != nil {
+		r.Event(obj, "Warning", "CurrentResourceParsing", err.Error())
+		resourceCondition.Status = metav1.ConditionFalse
+		meta.SetStatusCondition(&status.Conditions, resourceCondition)
+		obj.SetStatus(status.WithState(StateError).WithErr(err))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	if !meta.IsStatusConditionTrue(status.Conditions, resourceCondition.Type) {
```

Comment:
I think I really lost here, in which place we changed this `resourceCondition.status` to `true` so that it will not enter into this if block? seems to me this if condition will always return true.

Reply:
In line 192 we set it for the first time so the flow will be:

1. Reach 190 for the first, enter IF flow (as condition is still false), then change condition to true and exit
2. Reach 190 every consecutive run, see that it is either still true (from step 1) or it was reset (because of a failure in rendering during the regular consistency checks), then set it to true again
3. Be happy that now the resourceCondition is always true if the run succeeded :)

Reply:
ah, I have a wrong impression, I thought the `resourceCondition.Status` is always initialized to to false in the beginning, and only set to true inside this block. but actually this is just used to compare with status.Conditions which from actually resources, the status there is of cause different based on resources status.